doc/psidk/src/concepts/disk-storage.md

```diff
@@ -0,0 +1,163 @@
+
+When starting up a new infrastructure provider, the node will peer all IPFS data from the IPFS network. At minimum, therefore, the user running the infrastructure provider node must be able to pay for the bandwidth necessary to transfer all pinned files.
+
+### Publishing upload rates
+
+Every IP must publish their current rate that they charge for upload bandwidth. If they fail to set a rate, users will be able to consume the infrastructure provider's network bandwidth for free. This rate is then used to enforce a maximum billable amount after an upload is performed. Furthermore, as infrastructure providers adjust their rates, browser clients can automatically select the cheapest upload service providers to effectively achieve decentralized load balancing. <span style="color:red"> TODO - However, don't we have the same issue I described for download? An IP could set an excessively large upload rate to effectively outsource their upload responsibility to others, centralizing the service provision. If this was a shared fee instead, it would also simplify calculating the bond amount for third-party IPFS clients.</span>
+
+### Infrastructure provider properties
+
+The message that indicates that a user should be billed/refunded for an attempt to perform an upload or download will only be accepted if it comes from another active infrastructure provider. Therefore, if an infrastructure provider exits the active set partway through a user-upload, they will be unable to charge the user. As long as the IP has a bond sufficient to cover the file upload bandwidth, other producers will still peer the file from it and the file could still be fully uploaded. In this case, the user was able to upload the file for free. This is not a problem, since there's no way for the user to game this possibility to consistently achieve free uploads. If the bond of the former active infrastructure provider is insufficient to cover the bandwidth, the file upload will have failed to propagate on the network, but the user also will not have incurred any cost, so they must simply try the upload again with an active IP.
+
+### Legality
+
+An active IP with a valid file (the file hash matches its declared CID) must keep a file pinned, regardless of origin. <span style="color:red"> Are there legal issues here? An IP may require the right to blacklist certain content.</span>
```

Comment:
The probability of this being used to store illegal content seems higher to me than a blockchain that only has a concept of chain-storage, so we may want to be extra careful here. Do you recall why we need this guarantee? If an IP stops pinning a file (even if it matches a CID), then according to the "Billing details - Download" section the following should happen:

> If the CID of the file does exist in an irreversible block, but the node does not have the CID, it may simply return a 404 error code to the HTTP request.

That seems fine to me, and we can rely on L2 to kick out IPs who are failing to store legitimate data, and then IPs can feel slightly more protected legally in case they really need to blacklist content that someone uploaded to them.